### ISSUE TITLE
SweepFormula: Preserve channel information

### ISSUE THREAD

**Issue by t-b** [AllenInstitute/MIES#352]
Can we start color-coding the SF plot traces by headstage (in this PR)?

_Originally posted by @timjarsky in https://github.com/AllenInstitute/MIES/pull/341#issuecomment-549095073_

**Comment by ukos-git** [comment]
### current behavior

Currently, channel information is only available for inputs to the `data()` function like  `channels(AD0,AD9)`.

This currently creates a data wave of the following layout:

```
ROWS: data
COLS: sweep number grouping
LAYERS: channels
```

an input of `channels(AD0,AD9)` creates 2 items in `LAYERS`. an input of `channels(AD)` creates 1 item in LAYERS. On the plotting level we therefore do not have the channel information available for inputs like `channels(AD)`.

### preserve channel information

A clean solution to this is to create channel information for an input like `channels(AD)` to evaluate to the available channels. This can only be done in the  `SF_GetSweepForFormula(graph, range[0], range[1], channels[i][0], channels[i][1], sweeps)` which is the wave getter for the data() function. This function needs to handle the complete channels wave and build up a 3d-wave.

If AD consists of `AD0,AD9`, inputs like `channels(AD, AD0)` will then evaluate to a non-unique channel array `AD0,AD9,AD0` alignment in the LAYERS dimension.

### adopt statistical functions

Currently inputs that have more than one item in the LAYERS dimension (3d-waves) do not work with `min,max,mean` etc. as these functions all remove one dimension from the data structure. They act on the ROWS dimension and the dimension is removed afterwards.

I have added asserts for acting on LAYERED waves as it was not clear what to calculate from these waves. I now get what the minimum of a 3dimensional wave should produce and we can implement it. This needs to be secured through unit tests with a proper MWE for a 3-dimensional wave.

### color

Finally, if the last dimension is preserved by all functions and contains the channel information, color information can be added upon plotting.

### trace naming

I would also recommend adding dimension labels for the columns and layers so that traces can get named  with sweep number and channel type/number.

## TLDR;

Coloring the traces requires to preserve channel information in the last dimension.

**Comment by timjarsky** [comment]
@t-b @ukos-git will the SF wave have a fixed number of layers with specific layer/channel associations?

**Comment by t-b** [comment]
No only layers with data fill be added.

**Comment by ukos-git** [comment]
A fixed number of layers would mean that memory/storage will quickly be filled since empty data also uses up storage. Additionally, plotting data from channel1 vs data from channel2 would not be possible with a fixed channel layout.

A dimension label holding the channel number will allow identifying which channel exactly it is.

**Comment by t-b** [comment]
Note to self: Discuss how the new approach would then support different sampling intervals.

**Comment by ukos-git** [comment]
If by *sampling interval* you refer to the ROW layout, then I think this should not be discussed in this issue.
